[core.logging] Add RewriteAppender for filtering LogMeta.

lukeelmers opened the pull request:
Closes https://github.com/elastic/kibana/issues/57547

### Summary

Implements a `RewriteAppender` for the purposes of filtering / manipulating a `LogRecord`, as [described in the original issue](https://github.com/elastic/kibana/issues/57547#issuecomment-757623248). The approach here is based on log4j2's [RewriteAppender](http://logging.apache.org/log4j/2.x/manual/appenders.html#RewriteAppender).

Each rewrite appender uses a designated `policy` which describes what part of the `LogRecord` to manipulate. This PR starts with a single `meta` policy for modifying `LogMeta`. The policy has a `mode` which can either be `remove` or `update`, and takes in a list of paths in the `LogMeta` object which are to be modified.

For the time being we are intentionally keeping this undocumented in the public docs in case we choose to make future changes, as the legacy filtering behavior was also undocumented.

### Usage
Here's a full example configuration which logs both to the console and a log file, censoring the `cookie` request header from the http response logs:
```yaml
logging:
  appenders:
    console:
      type: console
      layout:
        type: pattern
        highlight: true
        pattern: "[%date][%level][%logger] %message %meta"
    file:
      type: file
      fileName: ./kibana.log
      layout:
        type: json
    censor:
      type: rewrite
      appenders: [console, file]
      policy:
        type: meta
        mode: update
        properties:
          - path: "http.request.headers.cookie"
            value: "[REDACTED]"
  loggers:
    - name: http.server.response
      appenders: [censor]
      level: debug
```

The `remove` mode does not take a value as it simply deletes the provided path:
```yaml
logging:
  appenders:
    censor:
      type: rewrite
      appenders: [console]
      policy:
        type: meta
        mode: remove
        properties:
          - path: "some.path.to.delete"
```

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107104/)
* Commit: 4372efa09979cfb2f5343ecab701721226a6023d
* This comment will update when the build is complete

### Failed CI Steps
* [Check Doc API Changes](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107104/execution/node/410/log)

### History
* :broken_heart: [Build #106838](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/106838/) failed 68e0019020124b7588a1babbeb350bbfb5d2b73c

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/106838/","number":"106838","commit":"68e0019020124b7588a1babbeb350bbfb5d2b73c"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107104/","number":"107104","commit":"4372efa09979cfb2f5343ecab701721226a6023d"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107129/)
* Commit: 0d7b5c93eadadc27b92ab8100ce284ffdf6c8111
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Integration Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107129/execution/node/458/log)

### History
* :broken_heart: [Build #107104](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107104/) failed 4372efa09979cfb2f5343ecab701721226a6023d
* :broken_heart: [Build #106838](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/106838/) failed 68e0019020124b7588a1babbeb350bbfb5d2b73c

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/106838/","number":"106838","commit":"68e0019020124b7588a1babbeb350bbfb5d2b73c"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107104/","number":"107104","commit":"4372efa09979cfb2f5343ecab701721226a6023d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107129/","number":"107129","commit":"0d7b5c93eadadc27b92ab8100ce284ffdf6c8111"}]}
PIPELINE-->

elasticmachine commented:
Pinging @elastic/kibana-core (Team:Core)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107140/)
* Commit: ae5facc750098f477a76d3240ed84a3af65d1662

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #107129](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107129/) failed 0d7b5c93eadadc27b92ab8100ce284ffdf6c8111
* :broken_heart: [Build #107104](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107104/) failed 4372efa09979cfb2f5343ecab701721226a6023d
* :broken_heart: [Build #106838](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/106838/) failed 68e0019020124b7588a1babbeb350bbfb5d2b73c

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/106838/","number":"106838","commit":"68e0019020124b7588a1babbeb350bbfb5d2b73c"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107104/","number":"107104","commit":"4372efa09979cfb2f5343ecab701721226a6023d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107129/","number":"107129","commit":"0d7b5c93eadadc27b92ab8100ce284ffdf6c8111"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107140/","number":"107140","commit":"ae5facc750098f477a76d3240ed84a3af65d1662"}]}
PIPELINE-->

TinaHeiligers commented:
@lukeelmers 

> For the time being we are intentionally keeping this undocumented in case we choose to make future changes, as the legacy filtering behavior was also undocumented.

While I understand that the initial implementation is pretty much a MVP, IMHO, I must disagree with not documenting the `RewriteAppender` _in at least the dev docs_. Being able to censor/filter sensitive data is a powerful feature, even at an MVP stage and I envisage it causing a few headaches down the line when folks are using logs for debugging.

lukeelmers commented:
@TinaHeiligers I'm perfectly comfortable documenting this for devs -- I think the only reason we previously decided to make this internal was because the legacy filtering is also undocumented and we wanted to reserve the right to make changes in the future (or decide whether we want to support this functionality long-term at all).

Since we don't yet have proper docs on the new logging system, I will just add a mention to the `README` for now and clarify that it's experimental/internal.

restrry commented on packages/kbn-logging/src/appenders.ts:
```diff
@@ -15,6 +15,7 @@ import { LogRecord } from './log_record';
  */
 export interface Appender {
   append(record: LogRecord): void;
+  update(config: { appenders?: Map<string, Appender> }): void;
```
or appender might be passed from outside. I haven't checked out how it works, but the interface in java looks like it is  https://github.com/apache/log4j/blob/trunk/src/main/java/org/apache/log4j/rewrite/RewriteAppender.java#L74
```js
const appenderNames = appender.getAppenders?();
if(appenderNames) {
  appenderNames.forEach(name => appender.setAppender(this.appender.get(name))
}
```

lukeelmers replied:
@restrry It looks like log4j has the concept of an [attachable appender](https://github.com/apache/log4j/blob/7be00eed88152dd011a619e8bae5a631235c3f4c/src/main/java/org/apache/log4j/spi/AppenderAttachable.java) internally, and they will take the appender refs from the configuration and [only attach those appenders to the rewrite appender](https://github.com/apache/log4j/blob/7be00eed88152dd011a619e8bae5a631235c3f4c/src/main/java/org/apache/log4j/xml/DOMConfigurator.java#L278-L284).

We could try to take a similar approach:
- Create `AttachableAppender` interface which extends from `Appender`, and have `RewriteAppender` implement that.
- `AttachableAppender` would need to have (at least) `addAppender`. I suppose it would also need a way to expose the required `appenderRefs` (otherwise we'd need to infer that from the computed config)
- Then instead of the lifecycle methods, in the logging system we could do something like:

```ts
    for (const [appenderKey, appenderConfig] of computedConfig.appenders) {
      this.appenders.set(appenderKey, Appenders.create(appenderConfig));
    }

    for (const [, appender] of this.appenders) {
      if (appender.isAttachable) {
        appender.appenderRefs.forEach(ref => appender.addAppender(this.appenders.get(ref));
      }
    }
```

WDYT? It would be nice to have the attached appenders scoped to only what is required and not need to mess with the `update` lifecycle, but OTOH it requires introducing the new concept of `AttachableAppender`

restrry replied:
>  It would be nice to have the attached appenders scoped to only what is required and not need to mess with the update lifecycle, but OTOH it requires introducing the new concept of AttachableAppender

What if we use duck-typing? Can get by with checking that `appender` has `addAppender` method?

lukeelmers replied:
> What if we use duck-typing? Can get by with checking that appender has addAppender method?

Yep, duck-typing is an option too, though I think it'd still look similar to my example above, even if we don't have an official `AttachableAppender` interface.

The simplest approach would be to check if it has `addAppender`, and then add all appenders to it.

If we want to scope it to the required appender references, we could expose `appender.appenderRefs` or similar (as in example above), and use that list to only call `addAppender` with the required refs.

lukeelmers replied:
I ended up duck-typing with `appenderRefs` and `addAppender`, though I did still add the methods as optional items in the appender interface. LMK what you think.

lukeelmers commented on src/core/server/logging/logging_config.ts:
```diff
@@ -153,7 +167,24 @@ export class LoggingConfig {
       } as AppenderConfigType,
     ],
     [
+      // By default, we transparently rewrite `LogMeta` sent to the
+      // console appender to remove potentially sensitive keys.
+      // This can be overridden in a logger's configuration.
       'console',
+      {
+        type: 'rewrite',
+        appenders: ['stdout'],
+        policy: {
+          type: 'meta',
+          mode: 'update',
+          properties: SENSITIVE_META_PATHS.map((path) => ({ path, value: REDACTED_META_TEXT })),
+        },
+      } as AppenderConfigType,
```
> Or filter out sensitive properties when creating meta object

@restrry Yeah, that's the approach we take now, the only issue is that it can't be disabled, and IIRC this was one of the reasons we wanted a rewrite appender to begin with, for the case where we have a reason to allow sensitive headers through for support/debugging purposes.

That's why the only alternative that comes to mind is a new config option to disable filters, which would then be used inside the http response logger `getResponseLog`. But in the spirit of having One Way Of Doing Things (TM), it'd be nice to use the existing config if we can.

Actually, as I think about it, this is an issue that would possibly be solved by [our discussion around having logger-specific default appender configurations](https://github.com/elastic/kibana/issues/90457).

> How it works now in the Legacy platform? Does it always censor the sensitive content?

It filters them out by default (just `authorization` and `cookie`), but you can disable that using the (undocumented) `logging.filter` settings - e.g. `logging.filter.cookie=none` to remove filter from that header in hapi/good. 
https://github.com/elastic/kibana/blob/72670ebc77442a54869f94e7bea2f1f8f3499461/packages/kbn-legacy-logging/src/get_logging_config.ts#L53-L60

restrry replied:
> Yeah, that's the approach we take now, the only issue is that it can't be disabled, and IIRC this was one of the reasons we wanted a rewrite appender to begin with, for the case where we have a reason to allow sensitive headers through for support/debugging purposes.

I'd say it's nice to have. It is better to not add such a feature than to add it but with the possibility of sensitive data leakage due to incorrect configuration: 
```
 While this ensures nothing sensitive is logged with a default configuration,
the protection is gone as soon as you customize your config to use something other than the console appender. 
```

> Actually, as I think about it, this is an issue that would possibly be solved by our discussion around having logger-specific default appender configurations.

Are you implying that we can hide the Meta object from the logs?  The `pattern` does not apply to `json` layout.

lukeelmers replied:
> I'd say it's nice to have. It is better to not add such a feature than to add it but with the possibility of sensitive data leakage due to incorrect configuration

Fair point, so do you think we should:
1. Keep the hardcoded filtering of sensitive header fields, meaning we can't turn those off for debugging purposes
2. Go with option 1 but also introduce a new config flag to disable this for debugging purposes (e.g. `logging.filterSensitiveHttpHeaders` or whatever). The flag would default to `true` but could be disabled if someone wanted to use a custom rewrite appender.
3. Some other option I'm not thinking of?

Ideally I'd like to not introduce another config flag, but I'm also unsure of how we could otherwise make it possible to disable the filtering.

> Are you implying that we can hide the Meta object from the logs?

No, sorry that was a bit vague. I meant if we went the route of having predefined default appender configurations on a per-logger basis (similar to the ES config file), then this would be another use case for those and would solve this problem. But if we go the simpler route of removing the `meta` from the `pattern` by default, that won't help us in this case.

restrry replied:
> Keep the hardcoded filtering of sensitive header fields, meaning we can't turn those off for debugging purposes

As for me, it's an acceptable trade-off. Any other opinions? @joshdover @pgayvallet

lukeelmers replied:
Yeah the more I think about it, the hardcoded approach feels like the safest option for the time being. And then if we were to consider an enhancement in 8.x to support logger-specific appender defaults as discussed in https://github.com/elastic/kibana/issues/90457, it would allow us to move these settings there anyway.

lukeelmers commented on src/core/server/logging/appenders/rewrite/policies/meta/meta_policy.ts:
```diff
@@ -0,0 +1,103 @@
+      if (has(record, `${META_PATH}.${path}`)) {
+        continue; // don't overwrite properties which already exist
+      }
+      set(record, `${META_PATH}.${path}`, value);
```
> Swap the leaf object only?

But that will still mutate references to the `LogMeta` object unless I deeply clone it... so I guess the question is whether avoiding mutating is worth the perf hit of calling `cloneDeep` on any `LogMeta` which needs to be filtered

restrry replied:
> so I guess the question is whether avoiding mutating is worth the perf hit of calling cloneDeep on any LogMeta which needs to be filtered

I'm torn. In general, it's a good practice, OTOH it definitely will cause perf penalty. Maybe  passing a copy of headers is enough https://github.com/elastic/kibana/blob/8b04961adaf8c728059209253778cfa1abb589ba/src/core/server/http/logging/get_response_log.ts#L48

lukeelmers replied:
> Maybe passing a copy of headers is enough

This seems like a good balance between the two options for me, and the perf hit is much smaller as it only impacts a few pieces of the http response logs -- whereas a deep clone in the RewriteAppender would, with the current default config, impact every single log event.

lukeelmers commented on src/core/server/logging/appenders/rewrite/policies/index.ts:
```diff
@@ -0,0 +1,31 @@
+
+export const createRewritePolicy = (config: RewritePolicyConfig): RewritePolicy => {
+  switch (config.type) {
+    case 'meta':
```
> Do we call so because we apply a policy to Meta object only?

Yeah exactly, they have a `map` and `properties` rewrite policy, so I called this `meta` as it is only scoped to the `LogMeta`. With this pattern, in the future we could have something like a `message` or `level` policy that will rewrite those items specifically.

lukeelmers commented on src/core/server/logging/appenders/rewrite/policies/index.ts:
```diff
@@ -0,0 +1,31 @@
+ */
+export type RewritePolicyConfig = MetaRewritePolicyConfig;
+
+export const rewritePolicyConfigSchema = schema.oneOf([metaRewritePolicyConfigSchema]);
```
👍  SGTM, some of this boilerplate was copied from the rolling file appender (which has multiple policies), so I'll remove for now as we don't need it yet.

lukeelmers commented on src/core/server/logging/appenders/rewrite/policies/meta/meta_policy.ts:
```diff
@@ -0,0 +1,103 @@
+   * The 'mode' specifies what action to perform on the
+   * specified properties.
+   *
+   * @remarks
```
No particular reason; I can move it to the summary

lukeelmers commented on src/core/server/logging/appenders/rewrite/policies/meta/meta_policy.ts:
```diff
@@ -0,0 +1,103 @@
+
+  private add(record: LogRecord) {
+    for (const { path, value } of this.config.properties) {
+      if (has(record, `${META_PATH}.${path}`)) {
```
Yeah that's what I originally had, I can switch it back

lukeelmers commented on src/core/server/logging/appenders/rewrite/rewrite_appender.ts:
```diff
@@ -0,0 +1,92 @@
+            'Be sure `appender.update()` is called before `appender.append()`.'
+        );
+      }
+      appender.append(transformedRecord);
```
Ah thank you for reminding me! I had been meaning to add a check to ensure a circular reference wasn't possible. Will sort that out.

lukeelmers replied:
Added a check to `logging_system`, however after I did this I thought it could alternately be rewritten in a custom `validate` on the config schema, LMK if you feel strongly one way or the other.

The advantage of doing it in the logging system is that we are only duck-typing "attachable appenders" in one place, whereas if we try to also check for any appender that has a top-level `appenders` property in the config schema, we're now duck-typing these appenders based on two different properties in two different places.

restrry replied:
> Added a check to logging_system, however after I did this I thought it could alternately be rewritten in a custom validate on the config schema, LMK if you feel strongly one way or the other.

Works for me, `applyBaseConfig` is called once per config update

TinaHeiligers commented on src/core/server/logging/appenders/rewrite/policies/meta/meta_policy.ts:
```diff
@@ -0,0 +1,103 @@
+    schema.object({
+      path: schema.string(),
+      value: schema.maybe(
+        schema.nullable(schema.oneOf([schema.string(), schema.number(), schema.boolean()]))
```
What would happen if we used changed the value at a path to something other than the original type for that value?
e.g:
```
	censor:
      kind: rewrite
      appenders: [console, file]
      policy:
        type: meta
        mode: update
        properties:
          - path: "http.request.headers.cookie"
            value: true
```
It's a curiosity question more than a code comment

lukeelmers replied:
> What would happen if we used changed the value at a path to something other than the original type for that value?

In the current implementation it would overwrite with whatever value you provided - at runtime there is no comparison between the existing value type with the replacement type.

TinaHeiligers replied:
I thought so, thanks. I'm assuming this isn't a problem and that folks indexing logs and parsing them for whatever purpose would know what their doing and handle the type change on their side. It's not a concern for now, we'll reconsider the consequences down the line if/when we implement other policy types.

lukeelmers replied:
> It's not a concern for now, we'll reconsider the consequences down the line if/when we implement other policy types.

Yeah I think it's still a valid point though. The main issue is that there isn't any way to know if the types are compatible until runtime, so if we wanted to enforce this, we'd either need to attempt to cast types, or throw an error.

Neither seems like a very appealing option. In particular throwing an error would be problematic as the person configuring the filter doesn't have control over the logger that's being filtered - so an error could be thrown any time after Kibana has already started, leading to unexpected server crashes.

For that reason, I agree with your assessment above that it is probably best to wait and see if this becomes a concern in practice. TBH I had originally implemented this just as a string, and only later added the other types as they are still valid yaml. But I'd be willing to bet the vast majority of filtering use cases are for string values.

TinaHeiligers reviewed (commented):
I went through the initial implementation and it seems to make sense. I tried to pull the code and run it locally but keep hitting a config error on start up (probably local).

I'd really like to see something in the README on the final implementation though.

lukeelmers commented:
> I tried to pull the code and run it locally but keep hitting a config error on start up (probably local).

I'll update the README, but the example config I gave in the PR description _should_ be working without an error if you want to play around with it.

lukeelmers commented on docs/development/core/server/kibana-plugin-core-server.savedobjectsfindresult.sort.md:
```diff
@@ -25,7 +25,7 @@ const page1 = await savedObjectsClient.find({
   sortField: 'updated_at',
   sortOrder: 'asc',
-  pit,
+  pit: { id },
```
I guess it is a bug in the extractor, docs were updated separately here: https://github.com/elastic/kibana/pull/91826

TinaHeiligers commented:
> ... but the example config I gave in the PR description _should_ be working without an error if you want to play around with it.

A direct copy-paste throws the following:
```
/Users/christianeheiligers/Projects/kibana/packages/kbn-config-schema/target/out/types/literal_type.js:17
    handleError(type, { value, valids: [expectedValue] }) {
                        ^
TypeError: Cannot destructure property 'value' of 'undefined' as it is undefined.
```
Changing `kind` to `type` gets rid of that but also doesn't work directly:
```
/Users/christianeheiligers/Projects/kibana/packages/kbn-config-schema/target/out/types/type.js:51
            throw new errors_1.ValidationError(error, namespace);
                  ^
Error: [config validation of [logging].appenders.file]: types that failed validation:
- [config validation of [logging].0.type]: expected value to equal [console]
- [config validation of [logging].1.fileName]: expected value of type [string] but got [undefined]
- [config validation of [logging].2.type]: expected value to equal [legacy-appender]
- [config validation of [logging].3.type]: expected value to equal [rewrite]
- [config validation of [logging].4.type]: expected value to equal [rolling-file]
```
Removing the file appender solves that. What am I missing in the config?

lukeelmers commented:
Ah sorry @TinaHeiligers, I added that example before https://github.com/elastic/kibana/pull/90764 was merged and some of the properties were renamed. I've updated now, let me know if you hit any issues.

lukeelmers commented on src/core/server/logging/appenders/rewrite/rewrite_appender.ts:
```diff
@@ -0,0 +1,92 @@
+    this.appenders = appenders;
+
+    // Ensure config only contains valid appenders.
+    const unknownAppenderKey = this.config.appenders.find(
```
But if we use `every`, we would need to do another search to find the unknown appender key to include in the error message. With the current implementation (borrowed from `logging_config`), we can at least get that first unknown key to throw in the error, which might help for debugging purposes.

lukeelmers replied:
I ended up changing this anyway based on the shift from `update` to `addAppender`

lukeelmers commented on src/core/server/logging/logging_system.ts:
```diff
@@ -167,6 +167,13 @@ export class LoggingSystem implements LoggerFactory {
+    // the other configured appenders. This enables appenders to act as a sort
+    // of middleware and call `append` on each other if needed.
+    for (const [, appenderConfig] of this.appenders) {
+      appenderConfig.update({ appenders: this.appenders });
```
Added a description for `addAppender`

lukeelmers commented:
@elasticmachine merge upstream

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107910/)
* Commit: e9b9bcc413b3e53d4183b7008a07aed0cbda1064
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-91492/e9b9bcc413b3e53d4183b7008a07aed0cbda1064)
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107910/execution/node/465/log)

### History
* :green_heart: [Build #107140](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107140/) succeeded ae5facc750098f477a76d3240ed84a3af65d1662
* :broken_heart: [Build #107129](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107129/) failed 0d7b5c93eadadc27b92ab8100ce284ffdf6c8111
* :broken_heart: [Build #107104](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107104/) failed 4372efa09979cfb2f5343ecab701721226a6023d
* :broken_heart: [Build #106838](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/106838/) failed 68e0019020124b7588a1babbeb350bbfb5d2b73c

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @lukeelmers

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/106838/","number":"106838","commit":"68e0019020124b7588a1babbeb350bbfb5d2b73c"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107104/","number":"107104","commit":"4372efa09979cfb2f5343ecab701721226a6023d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107129/","number":"107129","commit":"0d7b5c93eadadc27b92ab8100ce284ffdf6c8111"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107140/","number":"107140","commit":"ae5facc750098f477a76d3240ed84a3af65d1662"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107910/","number":"107910","commit":"e9b9bcc413b3e53d4183b7008a07aed0cbda1064"}]}
PIPELINE-->

restrry commented on src/core/server/logging/logging_system.ts:
```diff
@@ -146,6 +146,26 @@ export class LoggingSystem implements LoggerFactory {
+      if (stack.includes(key)) {
+        throw new Error(`Circular appender reference detected: [${stack.join(' -> ')} -> ${key}]`);
+      }
+      stack.push(key);
```
Shouldn't `stack` remove a `key` once it is processed?

restrry commented on src/core/server/logging/appenders/rewrite/rewrite_appender.ts:
```diff
@@ -0,0 +1,100 @@
+   * As appenders cannot be attached to each other until they are created,
+   * the `addAppender` method is used to pass in a configured appender.
+   */
+  public addAppender(appenderRef: string, appender: DisposableAppender) {
```
`appender: Appender`? 
It doesn't call `dispose` on a passed appender.

TinaHeiligers commented on src/core/server/logging/appenders/rewrite/policies/policy.ts:
```diff
@@ -9,11 +9,11 @@
  */
 export interface RewritePolicy {
   /**
    * Transforms a {@link LogRecord} based on the policy's configuration.
```
```suggestion:
   * Rewrites a {@link LogRecord} based on the policy's configuration.
  ```

TinaHeiligers commented on src/core/server/logging/appenders/rewrite/rewrite_appender.ts:
```diff
@@ -52,7 +52,7 @@ export class RewriteAppender implements DisposableAppender {
    */
   public append(record: LogRecord) {
-    const transformedRecord = this.policy.transform(record);
+    const rewriteedRecord = this.policy.rewrite(record);
```
nit: `rewriteedRecord` -> `rewrittenRecord`

lukeelmers replied:
Oops, that was a messed up grep. Thanks for catching 🙂

TinaHeiligers commented on src/core/server/logging/appenders/rewrite/rewrite_appender.ts:
```diff
@@ -0,0 +1,100 @@
+
+  /**
+   * Appenders can be "attached" to this one so that the RewriteAppender
+   * is able to act as a sort of middleware by calling `append` other appenders.
```
```suggestion
   * is able to act as a sort of middleware by calling `append` on other appenders.
```

TinaHeiligers reviewed (commented):
In general, I think the `rewriteAppender` will be useful in the long term. Maybe we can keep the basic functionality from this PR with the modification of hard coding removing the sensitive data from http logs? I still don't feel too comfortable about making the global assumption that folks want to not log that data but we can address that later on if needed.

lukeelmers commented on src/core/server/logging/README.md:
```diff
@@ -264,6 +264,125 @@ The maximum number of files to keep. Once this number is reached, oldest files w
+
+Meta policies must specify one of three modes, which indicate which action
+to perform on the configured properties:
+- `add` creates a new property at the provided `path`, skipping properties which already exist.
```
Yes exactly, `add` creates a new property only if it doesn't already exist (to avoid unexpectedly overwriting a value).

Conversely, `update` changes a property only if it already exists (to avoid unexpectedly creating a new value).

TinaHeiligers commented:
👍
Christiane (Tina) Heiligers


On Fri, Feb 19, 2021 at 11:30 AM Luke Elmers <notifications@github.com>
wrote:

> *@lukeelmers* commented on this pull request.
> ------------------------------
>
> In src/core/server/logging/README.md
> <https://github.com/elastic/kibana/pull/91492#discussion_r579391236>:
>
> > +The `meta` rewrite policy can read and modify any data contained in the
> +`LogMeta` before passing it along to a destination appender.
> +
> +```yaml
> +logging:
> +  appenders:
> +    my-rewrite-appender:
> +      type: rewrite
> +      appenders: [console, file] # name of "destination" appender(s)
> +      policy:
> +        # ...
> +```
> +
> +Meta policies must specify one of three modes, which indicate which action
> +to perform on the configured properties:
> +- `add` creates a new property at the provided `path`, skipping properties which already exist.
>
> Yes exactly, add creates a new property only if it doesn't already exist
> (to avoid unexpectedly overwriting a value).
>
> Conversely, update changes a property only if it already exists (to avoid
> unexpectedly creating a new value).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/kibana/pull/91492#discussion_r579391236>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC23SSUNOBAH2YJTJ33G4NTS72U5JANCNFSM4XWSY7VA>
> .
>

lukeelmers commented on src/core/server/logging/appenders/rewrite/policies/meta/meta_policy.test.ts:
```diff
@@ -0,0 +1,228 @@
+  ) => new MetaRewritePolicy({ type: 'meta', mode, properties });
+
+  const createLogRecord = (meta: LogMeta = {}): LogRecord => ({
+    timestamp: new Date(),
```
👍  The current tests aren't making any assertions based on timestamp, however I'll go ahead and use a fixed timestamp as a precautionary measure.

lukeelmers commented:
> Maybe we can keep the basic functionality from this PR with the modification of hard coding removing the sensitive data from http logs?

Yep, my plan is to add the hardcoded logic back in, was just waiting to see if any others wanted to chime in on that thread with any strong reasons not to do so. It sounds like so far we're all on the same page though.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108086/)
* Commit: a6632ee06b46a22f2267373b50df0e93f83b12dc
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-91492/a6632ee06b46a22f2267373b50df0e93f83b12dc)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #107910](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107910/) failed e9b9bcc413b3e53d4183b7008a07aed0cbda1064
* :green_heart: [Build #107140](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107140/) succeeded ae5facc750098f477a76d3240ed84a3af65d1662
* :broken_heart: [Build #107129](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107129/) failed 0d7b5c93eadadc27b92ab8100ce284ffdf6c8111
* :broken_heart: [Build #107104](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107104/) failed 4372efa09979cfb2f5343ecab701721226a6023d
* :broken_heart: [Build #106838](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/106838/) failed 68e0019020124b7588a1babbeb350bbfb5d2b73c

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @lukeelmers

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/106838/","number":"106838","commit":"68e0019020124b7588a1babbeb350bbfb5d2b73c"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107104/","number":"107104","commit":"4372efa09979cfb2f5343ecab701721226a6023d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107129/","number":"107129","commit":"0d7b5c93eadadc27b92ab8100ce284ffdf6c8111"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107140/","number":"107140","commit":"ae5facc750098f477a76d3240ed84a3af65d1662"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107910/","number":"107910","commit":"e9b9bcc413b3e53d4183b7008a07aed0cbda1064"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108086/","number":"108086","commit":"a6632ee06b46a22f2267373b50df0e93f83b12dc"}]}
PIPELINE-->

lukeelmers commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108243/)
* Commit: e0f4c583954c6a382a0417391ce581ec2846630f
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-91492/e0f4c583954c6a382a0417391ce581ec2846630f)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #108086](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108086/) succeeded a6632ee06b46a22f2267373b50df0e93f83b12dc
* :broken_heart: [Build #107910](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107910/) failed e9b9bcc413b3e53d4183b7008a07aed0cbda1064
* :green_heart: [Build #107140](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107140/) succeeded ae5facc750098f477a76d3240ed84a3af65d1662
* :broken_heart: [Build #107129](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107129/) failed 0d7b5c93eadadc27b92ab8100ce284ffdf6c8111
* :broken_heart: [Build #107104](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107104/) failed 4372efa09979cfb2f5343ecab701721226a6023d

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @lukeelmers

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107104/","number":"107104","commit":"4372efa09979cfb2f5343ecab701721226a6023d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107129/","number":"107129","commit":"0d7b5c93eadadc27b92ab8100ce284ffdf6c8111"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107140/","number":"107140","commit":"ae5facc750098f477a76d3240ed84a3af65d1662"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107910/","number":"107910","commit":"e9b9bcc413b3e53d4183b7008a07aed0cbda1064"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108086/","number":"108086","commit":"a6632ee06b46a22f2267373b50df0e93f83b12dc"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108243/","number":"108243","commit":"e0f4c583954c6a382a0417391ce581ec2846630f"}]}
PIPELINE-->

pgayvallet commented on src/core/server/http/logging/get_response_log.ts:
```diff
@@ -45,7 +45,12 @@ export function getEcsResponseLog(request: Request, log: Logger): LogMeta {
   const status_code = isBoom(response) ? response.output.statusCode : response.statusCode;
-  const responseHeaders = isBoom(response) ? response.output.headers : response.headers;
+
+  // shallow clone the headers so they are not mutated if filtered by a RewriteAppender
+  const requestHeaders = { ...request.headers };
+  const responseHeaders = isBoom(response)
+    ? { ...response.output.headers }
+    : { ...response.headers };
```
NIT: is shallow clone sufficient here? for array headers, the appender could potentially still mutates the values. Not sure if this is a real risk in practice though?

lukeelmers replied:
Yeah that's a good point, @restrry and I were debating the tradeoffs of perf vs preventing mutation as this is a pretty hot code path.

It seems odd that someone would want to overwrite only one specific array item as they'd need to do it by index - the appender config would need to be `{ path: 'http.request.headers[0]', value: 'whatever' }`. But of course it is possible so I suppose it doesn't hurt to spread out the array here.

lukeelmers replied:
Updated to prevent mutating array headers, just to be safe

pgayvallet commented on src/core/server/logging/appenders/rewrite/policies/meta/meta_policy.ts:
```diff
@@ -0,0 +1,98 @@
+   * @remarks
+   * Each provided 'path' is relative to the record's {@link LogMeta}.
+   * For the 'remove' mode, no 'value' is provided.
+   */
+  properties: Array<{ path: string; value?: string | number | boolean | null }>;
```
NIT/optional: I think I would extract the inline type.

pgayvallet commented on src/core/server/logging/README.md:
```diff
@@ -264,6 +264,127 @@ The maximum number of files to keep. Once this number is reached, oldest files w
+          - path: "http.request.headers.authorization"
+          - path: "http.request.headers.cookie"
+          - path: "http.request.headers.set-cookie"
+    add-stuff:
+      type: rewrite
+      appenders: [console] # output to console
+      policy:
+        type: meta
+        mode: add
+        properties:
+          - path: "hello"
+            value: "world" # creates { hello: 'world' } at the LogMeta root
```
is the `add` mode without any value interpolation really relevant?

lukeelmers replied:
Fair question, I mostly included an `add` mode as [log4j has one](https://logging.apache.org/log4j/2.x/manual/appenders.html#RewritePolicy), and it didn't really require extra work to support.

But admittedly update/remove seem to be the more useful modes. (FWIW the example usage log4j gives for `add` is injecting a static product key). Not opposed to removing it though, if we don't feel it's worth keeping. Thoughts?

TinaHeiligers replied:
If we don't have a need to support an `add` mode, maybe we don't need it right now.
My vote is that we don't have a direct need for it and shouldn't include it right now.

lukeelmers replied:
Removed

pgayvallet commented on src/core/server/logging/appenders/rewrite/rewrite_appender.ts:
```diff
@@ -0,0 +1,100 @@
+   * As appenders cannot be attached to each other until they are created,
+   * the `addAppender` method is used to pass in a configured appender.
+   */
+  public addAppender(appenderRef: string, appender: Appender) {
+    if (!this.appenderRefs.includes(appenderRef)) {
+      throw new Error(
+        `addAppender was called with an appender key that is missing from the appenderRefs: "${appenderRef}".`
+      );
+    }
+
```
Too early for it probably, but we might want a `MiddlewareAppender` base / abstract class at some point in the future if we're going to add additional 'middlewares' to handle this `addAppender` logic, and most of the implementation of `append` (the nested appender looping)

lukeelmers replied:
Log4j has something like this internally called `AttachableAppender` - Basically an appender that you can attach other appenders to. I thought about doing the same thing here but wasn't sure if it was worth introducing a new concept since we only have one concrete use case for it ATM.

Open to adding it though; WDYT?

restrry replied:
Let's wait for more cases before introducing a new abstraction.

pgayvallet commented on src/core/server/logging/logging_system.ts:
```diff
@@ -167,6 +187,23 @@ export class LoggingSystem implements LoggerFactory {
+      }
+      for (const ref of appender.appenderRefs) {
+        const foundAppender = this.getAppenderByRef(ref);
+        if (!foundAppender) {
+          throw new Error(`Appender config contains unknown appender key "${ref}".`);
+        }
```
NIT:

```ts
// const [id, appender] of this.appenders

throw new Error(`Appender "${id}" config contains unknown appender key "${ref}".`);
```

pgayvallet reviewed (commented):
Few NITS, overall looking great

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108584/)
* Commit: dcee4b9388359465fbb1018452bb18f791d60f22
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-91492/dcee4b9388359465fbb1018452bb18f791d60f22)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #108243](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108243/) succeeded e0f4c583954c6a382a0417391ce581ec2846630f
* :green_heart: [Build #108086](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108086/) succeeded a6632ee06b46a22f2267373b50df0e93f83b12dc
* :broken_heart: [Build #107910](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107910/) failed e9b9bcc413b3e53d4183b7008a07aed0cbda1064
* :green_heart: [Build #107140](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107140/) succeeded ae5facc750098f477a76d3240ed84a3af65d1662
* :broken_heart: [Build #107129](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107129/) failed 0d7b5c93eadadc27b92ab8100ce284ffdf6c8111

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @lukeelmers

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107129/","number":"107129","commit":"0d7b5c93eadadc27b92ab8100ce284ffdf6c8111"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107140/","number":"107140","commit":"ae5facc750098f477a76d3240ed84a3af65d1662"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107910/","number":"107910","commit":"e9b9bcc413b3e53d4183b7008a07aed0cbda1064"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108086/","number":"108086","commit":"a6632ee06b46a22f2267373b50df0e93f83b12dc"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108243/","number":"108243","commit":"e0f4c583954c6a382a0417391ce581ec2846630f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108584/","number":"108584","commit":"dcee4b9388359465fbb1018452bb18f791d60f22"}]}
PIPELINE-->

restrry commented on src/core/server/http/logging/get_response_log.ts:
```diff
@@ -18,14 +18,22 @@ const ECS_VERSION = '1.7.0';
-      result[key] = FORBIDDEN_HEADERS.includes(key) ? REDACTED_HEADER_TEXT : headers[key];
+      result[key] = redactSensitiveHeaders(
+        key,
+        Array.isArray(headers[key]) ? [...headers[key]] : headers[key]
```
It's good we enforce immutability, but we pay the extra cost when cloning the array. IMO we can assume that no-one mutates the array.

restrry commented on src/core/server/logging/logging_config.ts:
```diff
@@ -153,7 +168,26 @@ export class LoggingConfig {
+        policy: {
+          type: 'meta',
+          mode: 'update',
+          properties: SENSITIVE_META_PATHS.map((path) => ({ path, value: REDACTED_META_TEXT })),
```
Can be removed now since you filter the sensitive headers during `LogRecord` creation?

lukeelmers replied:
It depends on whether we want to try to filter any other ECS log meta by default or not. But TBH I don't know who else would be logging request/response headers in this way, so it's probably ok to remove.

restrry replied:
> I don't know who else would be logging request/response headers

Even if another plugin logged them, it would be logged under a different context name, so the filtering mechanism doesn't filter them out.

lukeelmers replied:
> it would be logged under a different context name

Correct, I meant that the defaults applied in `logging_config` would apply to any other context name, unlike the hardcoded filter in the http response logs.

However as discussed in another thread here, this only would work until someone is no longer using the default logging config. So the cost of keeping this extra code in place may not be worth the benefit of filtering the headers in this very specific scenario (another plugin logging headers in ECS format when Kibana has a default logging config)

lukeelmers replied:
I went ahead and removed this for now. Perhaps best to wait and revisit if we decide to pursue https://github.com/elastic/kibana/issues/92082

lukeelmers commented:
@elasticmachine merge upstream

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108933/)
* Commit: 102d490fe14cc168563a03e376a2d5d49e107e3f
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-91492/102d490fe14cc168563a03e376a2d5d49e107e3f)
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108933/execution/node/498/log)
* [Execute xpack-kibana-ciGroup6](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108933/execution/node/758/log)

### History
* :green_heart: [Build #108584](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108584/) succeeded dcee4b9388359465fbb1018452bb18f791d60f22
* :green_heart: [Build #108243](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108243/) succeeded e0f4c583954c6a382a0417391ce581ec2846630f
* :green_heart: [Build #108086](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108086/) succeeded a6632ee06b46a22f2267373b50df0e93f83b12dc
* :broken_heart: [Build #107910](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107910/) failed e9b9bcc413b3e53d4183b7008a07aed0cbda1064
* :green_heart: [Build #107140](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107140/) succeeded ae5facc750098f477a76d3240ed84a3af65d1662

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @lukeelmers

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107140/","number":"107140","commit":"ae5facc750098f477a76d3240ed84a3af65d1662"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107910/","number":"107910","commit":"e9b9bcc413b3e53d4183b7008a07aed0cbda1064"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108086/","number":"108086","commit":"a6632ee06b46a22f2267373b50df0e93f83b12dc"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108243/","number":"108243","commit":"e0f4c583954c6a382a0417391ce581ec2846630f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108584/","number":"108584","commit":"dcee4b9388359465fbb1018452bb18f791d60f22"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108933/","number":"108933","commit":"102d490fe14cc168563a03e376a2d5d49e107e3f"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108932/)
* Commit: 3d2a945f142d96848247c2420ae79fbd3368a2db
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-91492/3d2a945f142d96848247c2420ae79fbd3368a2db)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #108933](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108933/) failed 102d490fe14cc168563a03e376a2d5d49e107e3f
* :green_heart: [Build #108584](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108584/) succeeded dcee4b9388359465fbb1018452bb18f791d60f22
* :green_heart: [Build #108243](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108243/) succeeded e0f4c583954c6a382a0417391ce581ec2846630f
* :green_heart: [Build #108086](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108086/) succeeded a6632ee06b46a22f2267373b50df0e93f83b12dc
* :broken_heart: [Build #107910](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107910/) failed e9b9bcc413b3e53d4183b7008a07aed0cbda1064

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @lukeelmers

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/107910/","number":"107910","commit":"e9b9bcc413b3e53d4183b7008a07aed0cbda1064"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108086/","number":"108086","commit":"a6632ee06b46a22f2267373b50df0e93f83b12dc"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108243/","number":"108243","commit":"e0f4c583954c6a382a0417391ce581ec2846630f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108584/","number":"108584","commit":"dcee4b9388359465fbb1018452bb18f791d60f22"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108933/","number":"108933","commit":"102d490fe14cc168563a03e376a2d5d49e107e3f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108932/","number":"108932","commit":"3d2a945f142d96848247c2420ae79fbd3368a2db"}]}
PIPELINE-->

lukeelmers commented:
@elasticmachine merge upstream

TinaHeiligers reviewed (approved):
This is a great addition to our new logging system and even though we're not using the implementation in the default config right now, it's going to be super useful when we get around to https://github.com/elastic/kibana/issues/92082.
LGTM!

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/109205/)
* Commit: fb23cb1672e71bc2c3775fe8c473ab2d6b5f20fa
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-91492/fb23cb1672e71bc2c3775fe8c473ab2d6b5f20fa)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #108933](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108933/) failed 102d490fe14cc168563a03e376a2d5d49e107e3f
* :green_heart: [Build #108932](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108932/) succeeded 3d2a945f142d96848247c2420ae79fbd3368a2db
* :green_heart: [Build #108584](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108584/) succeeded dcee4b9388359465fbb1018452bb18f791d60f22
* :green_heart: [Build #108243](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108243/) succeeded e0f4c583954c6a382a0417391ce581ec2846630f
* :green_heart: [Build #108086](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108086/) succeeded a6632ee06b46a22f2267373b50df0e93f83b12dc

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @lukeelmers

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108086/","number":"108086","commit":"a6632ee06b46a22f2267373b50df0e93f83b12dc"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108243/","number":"108243","commit":"e0f4c583954c6a382a0417391ce581ec2846630f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108584/","number":"108584","commit":"dcee4b9388359465fbb1018452bb18f791d60f22"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108932/","number":"108932","commit":"3d2a945f142d96848247c2420ae79fbd3368a2db"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/108933/","number":"108933","commit":"102d490fe14cc168563a03e376a2d5d49e107e3f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/109205/","number":"109205","commit":"fb23cb1672e71bc2c3775fe8c473ab2d6b5f20fa"}]}
PIPELINE-->

kibanamachine commented:
## 💚 Backport successful

✅ [7.x](https://github.com/elastic/kibana/pull/92724) / https://github.com/elastic/kibana/pull/92724

Successful backport PRs will be merged automatically after passing CI.
